File: prefix_sum/525_contiguous_array-solved/solution.ts
/*
 * 525. Contiguous Array
 * https://leetcode.com/problems/contiguous-array/
 *
 * Given a binary array nums, return the maximum length of a contiguous
 * subarray with an equal number of 0 and 1.
 *
 * Example 1: nums=[0,1] → 2
 * Example 2: nums=[0,1,0] → 2
 */

function findMaxLength(nums: number[]): number {
    let sum = 0
    const sumMap = new Map()
    let maxLen = 0
    sumMap.set(0, -1)
    for (let i = 0; i < nums.length; i++) {
        sum += nums[i] === 0 ? -1 : nums[i]
        if (sumMap.has(sum)) {
            maxLen = Math.max(maxLen, i - sumMap.get(sum))
        } else {
            sumMap.set(sum, i)
        }
    }
    return maxLen
};

File: prefix_sum/525_contiguous_array-solved/solution.py
"""
525. Contiguous Array
https://leetcode.com/problems/contiguous-array/

Given a binary array nums, return the maximum length of a contiguous subarray
with an equal number of 0 and 1.

Example 1:
    Input: nums = [0,1]
    Output: 2
    Explanation: [0, 1] has equal numbers of 0 and 1.

Example 2:
    Input: nums = [0,1,0]
    Output: 2
    Explanation: [0, 1] (or [1, 0]) is the longest with equal 0 and 1.

Constraints:
    1 <= nums.length <= 10^5
    nums[i] is either 0 or 1.
"""

from typing import List


class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
